airflow/models/taskinstance.py

```diff
@@ -2335,9 +2335,11 @@ def _record_task_map_for_downstreams(self, task: "Operator", value: Any, *, sess
             return
         if value is None:
             raise XComForMappingNotPushed()
-        for validator in validators:
-            validator(value)
-        assert isinstance(value, collections.abc.Collection)  # The validators type-guard this.
+        if not isinstance(value, (collections.abc.Sequence, dict)):
+            raise UnmappableXComTypePushed(value)
+        if isinstance(value, (bytes, str)):
+            raise UnmappableXComTypePushed(value)
+        assert isinstance(value, collections.abc.Collection)  # The isinstance() checks above guard this.
```

Comment:
```suggestion
        if TYPE_CHECKING:
            assert isinstance(value, collections.abc.Collection)  # The isinstance() checks above guard this.
```

Is this assert just for typing? If so, we should [only do it when TYPE_CHECKING](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#dont-use-asserts-outside-tests).